src/units/abilities.cpp

```diff
@@ -966,14 +989,232 @@ void attack_type::modified_attacks(bool is_backstab, unsigned & min_attacks,
 	int damage_value = dmg_effect.get_composite_value();
-	if ( combat_ability("damage", damage_value, is_backstab).second ) {
-		damage_value = combat_ability("damage", damage_value, is_backstab).first;
-	}
 	return damage_value;
 }
 
+//The following functions are intended to allow the use in combat of capacities
+//identical to special weapons and therefore to be able to use them on adjacent
+//units (abilities of type 'aura') or else on all types of weapons even if the
+//beneficiary unit does not have a corresponding weapon
+//(defense against ranged weapons abilities for a unit that only has melee attacks)
+
+//The two functions below are used to manage filters similar to those of
+//special weapons ([filter_self] is replaced by [filter_student])
+static bool ability_filter_fighter(const std::string& ability,
+				   const std::string & filter_attacker,
+				   const config& cfg, const map_location & loc,
+				   unit_const_ptr & u,
+				   unit_const_ptr & u2,
+				   const_attack_ptr weapon)
+{
+	
+	if (!loc.valid()){
+		return true;
+	}
+	const config &filter = cfg.child(filter_attacker);
+	if(!filter) {
+		return true;
+	}
+	
+	if (!u) {
+		return false;
+	}
+	
+	if ( const config & filter_weapon = filter.child("filter_weapon") ) {
+		if ( !weapon || !weapon->matches_filter(filter_weapon) )
+			return false;
+	}
+	
+	if (!u2) {
+		return unit_filter(vconfig(filter)).set_use_flat_tod(ability == "illuminates").matches(*u, loc);
+	}
+	return unit_filter(vconfig(filter)).set_use_flat_tod(ability == "illuminates").matches(*u, loc, *u2);
+}
+
+static bool ability_apply_filter(unit_const_ptr un, unit_const_ptr up, const std::string& ability, const config& cfg, const map_location& loc, const map_location& opp_loc, bool attacker, const_attack_ptr weapon, const_attack_ptr opp_weapon)
+{
+	bool filter_opponent = !ability_filter_fighter(ability, "filter_opponent", cfg, opp_loc, up, un, opp_weapon);
+	bool filter_student = !ability_filter_fighter(ability, "filter_student", cfg, loc, un, up, weapon);
+	bool filter_attacker = (attacker && !ability_filter_fighter(ability, "filter_attacker", cfg, loc, un, up, weapon)) || (!attacker && !ability_filter_fighter(ability, "filter_attacker", cfg, opp_loc, up, un, opp_weapon));
+	bool filter_defender = (!attacker && !ability_filter_fighter(ability, "filter_defender", cfg, loc,un, up, weapon)) || (attacker && !ability_filter_fighter(ability, "filter_defender", cfg, opp_loc, up, un, opp_weapon));
+	if(filter_student || filter_opponent || filter_attacker || filter_defender){
+		return true;
+	}
+	return false;
+}
+
+
+bool leadership_affects_self(const std::string& ability, const unit &un, const map_location& loc, bool attacker)
+{
+	unit_ability_list abil = un.get_abilities(ability, loc);
+	for(unit_ability_list::iterator i = abil.begin(); i != abil.end();) {
+		const std::string& apply_to = (*i->first)["apply_to"];
+		if(apply_to.empty() || apply_to == "both" || apply_to == "self") {
```

Comment:
is there a reason why you reimplemeted this instead of using special_affects_self ?